capstone/src/main/webapp/sheetsScript.js

```diff
@@ -62,29 +64,272 @@ async function gatherSheetsInput() {
+ * as the one being analyzed
+ */
+async function preprocessSheet(id, sheetNames) {
+  var requests = [];
+
+  for (const name of sheetNames) {
+    requests.push({ addSheet: { properties: { title: name }}});
+  }
+
+  requests.push({ deleteSheet: { sheetId: 0 }});
+
+  await gapi.client.sheets.spreadsheets.batchUpdate({
+    spreadsheetId: id,
+    resource: { requests }
+  });
+}
+
+/**
+ * Return a JSON that contains row-by-row Perspective analysis for
+ * every cell in a Google Sheet
+ */
+async function createSheetOutput(id, name, range, lang, requestedAttributes) {
+  const sheet = await getSpreadsheet(id, name, range);
+
+  if (sheet.values.length > 0) {
+    // Create a key as the top row
+    let body = [Array.from(requestedAttributes).sort()];
+    body[0].unshift('Comment');
+
+    // Create Perspective calls
+    const promises = [];
+    for (i = 0; i < sheet.values.length; i++) {
+      let row = sheet.values[i];
+      for (j = 0; j < row.length; j++) {
+        if (row[j] != '') {
+          promises.push(callPerspective(row[j], lang, requestedAttributes));
+          body.push([row[j]]);
+        }
+      }
+    }
+
+    // Evaluate Perspective calls
+    const resolvedResponses = await Promise.all(promises);
+    for (index = 0; index < promises.length; index++) {
+      if (resolvedResponses[index].attributeScores) {
+        // Create a new row with the comment and Perspective scores
+        for (const attribute of Object.entries(resolvedResponses[index].attributeScores).sort()) {
+          body[index + 1].push(attribute[1].summaryScore.value);
+        }
+      }
+    }
+
+    return body;
+  }
+  return 'No data found.';
+}
+
+/**
+ * Append body to the Sheet with an id of id
+ */
+async function appendDataToSheet(id, name, body) {
+  await gapi.client.sheets.spreadsheets.values.append({
+    spreadsheetId: id,
+    range: name + '!A1:YY',
+    valueInputOption: 'USER_ENTERED',
+    resource: { values: body }
+  });
+}
+
+/**
+ * Color code number values within the row and column range
+ * in the Sheet with corresponding spreadsheetId and sheetId
+ */
+async function addFormatting(spreadsheetId, sheetId, numRows, numCols) {
+  const myRange = {
+    sheetId: sheetId,
+    startRowIndex: 0,
+    endRowIndex: numRows,
+    startColumnIndex: 0,
+    endColumnIndex: numCols,
+  };
+  var requests = [{
+      addConditionalFormatRule: {
+        rule: {
+          ranges: [ myRange ],
+          gradientRule: {
+            minpoint: { color: { red: 0.345, green: 0.733, blue: 0.541, alpha: 1 }, type: 'NUMBER', value: '0' },
+            midpoint: { color: { red: 1, green: 0.839, blue: 0.4, alpha: 1 }, type: 'PERCENTILE', value: '50' },
+            maxpoint: { color: { red: 0.902, green: 0.486, blue: 0.451, alpha: 1}, type: 'NUMBER', value: '1' }
+          }
+        },
+        index: 0
+      }}];
+
+  await gapi.client.sheets.spreadsheets.batchUpdate({
+    spreadsheetId: spreadsheetId,
+    resource: { requests }
```

Comment:
nit: Is it possible to remove the brackets?

Reply:
Not that I know of - the API takes an object with an array of requests